File: core/src/ru/myitschool/dcrawler/story/quest/Quest.java
package ru.myitschool.dcrawler.story.quest;

import com.badlogic.gdx.Game;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import ru.myitschool.dcrawler.dungeon.Room;
import ru.myitschool.dcrawler.screens.FinalScreen;

import java.util.Random;

/**
 * Created by Voyager on 21.03.2018.
 */
public abstract class Quest {

    private boolean changeRoom;

    private int roomPos;
    private int deltaPos;
    private int roomsMax;

    private static Color exitColor = new Color(0xffcc2288);

    private Room room;

    public Quest(int roomPos, int deltaPos, int roomsMax, Room room) {
        this(roomPos, deltaPos, roomsMax);
        this.room = room;
        this.changeRoom = true;
    }

    public Quest(int roomPos, int deltaPos, int roomsMax) {
        this.roomPos = roomPos;
        this.deltaPos = deltaPos;
        if (roomsMax >= roomPos + deltaPos){
            this.roomsMax = roomsMax;
        } else {
            this.roomsMax = roomPos + deltaPos;
        }
        this.room = null;
        this.changeRoom = false;
    }

    public boolean isChangeRoom() {
        return changeRoom;
    }

    public Room getRoom() {
        return room;
    }

    public void setRoom(Room room) {
        this.room = room;
        changeRoom = room != null;
    }

    public int getRoomPos() {
        return roomPos + new Random().nextInt(deltaPos + 1);
    }

    public int getRoomsMax() {
        return roomsMax;
    }

    public static Color getExitColor() {
        return exitColor;
    }

    public Room exitRoomConstruction(Room room){
        if (isChangeRoom()){
            room = getRoom();
        }
        room.setMobs(false);
        room = modifyRoom(room);
        return room;
    }

    public abstract Room modifyRoom(Room room);

    public abstract void exitRoomOpened();

    public abstract void targetReached();

    public void complete(){
        ((Game) Gdx.app.getApplicationListener()).setScreen(new FinalScreen("Complete!"));
    }
}

File: core/src/com/wolg_vlad/dcrawler/story/quest/Quest.kt
package com.wolg_vlad.dcrawler.story.quest

import com.badlogic.gdx.Game
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.wolg_vlad.dcrawler.dungeon.Room
import com.wolg_vlad.dcrawler.ui.screens.FinalScreen
import com.wolg_vlad.dcrawler.utils.SeededRandom

/**
 * Created by Voyager on 21.03.2018.
 */
abstract class Quest(endRoom: Int, private val deltaPos: Int, roomsMax: Int) {
    val endRoom: Int
    val roomsMax: Int
    private var room: Room? //todo rework. Why should all of the quests have room?

    constructor(roomPos: Int, deltaPos: Int, roomsMax: Int, room: Room) : this(roomPos, deltaPos, roomsMax) {
        this.room = room
    }

    fun getRoom(): Room? {
        return room
    }

    fun setRoom(room: Room) {
        this.room = room
    }

    fun exitRoomConstruction(room: Room): Room {
        var room = room
        if (getRoom() != null) {
            room = getRoom() as Room
        }
        room!!.setMobs(false)
        room = modifyRoom(room)
        return room
    }

    abstract fun modifyRoom(room: Room): Room
    abstract fun exitRoomOpened()
    abstract fun targetReached()
    fun complete() {
        (Gdx.app.applicationListener as Game).screen = FinalScreen("Complete!")
    }

    companion object {
        val exitColor = Color(-0x33dd78)
    }

    init {
        if (roomsMax >= endRoom + deltaPos) {
            this.roomsMax = roomsMax
        } else {
            this.roomsMax = endRoom + deltaPos
        }
        this.endRoom = endRoom + SeededRandom.getInstance().nextInt(deltaPos + 1)
        room = null
    }
}